drkafka/src/main/java/com/pinterest/doctorkafka/KafkaCluster.java

```diff
@@ -245,38 +278,155 @@ public BrokerStats getLatestBrokerStats(int brokerId) {
 
     boolean success = true;
     Map<Integer, KafkaBroker> result = new HashMap<>();
-    List<KafkaBroker> unusableBrokers = new ArrayList<>();
+    Set<KafkaBroker> unusableBrokers = new HashSet<>();
+
     for (int oosBrokerId : oosReplica.outOfSyncBrokers) {
       // we will get the broker with the least network usage
-      KafkaBroker leastUsedBroker = brokerQueue.poll();
-      while (leastUsedBroker != null && oosReplica.replicaBrokers.contains(leastUsedBroker.id())) {
-        unusableBrokers.add(leastUsedBroker);
-        leastUsedBroker = brokerQueue.poll();
-      }
-      if (leastUsedBroker == null) {
-        LOG.error("Failed to find a usable broker for fixing {}:{}", oosReplica, oosBrokerId);
-        success = false;
-      } else {
-        LOG.info("LeastUsedBroker for replacing {} : {}", oosBrokerId, leastUsedBroker.id());
-        success &= leastUsedBroker.reserveInBoundBandwidth(topicPartition, inBoundReq);
-        if (preferredBroker == oosBrokerId) {
-          success &= leastUsedBroker.reserveOutBoundBandwidth(topicPartition, outBoundReq);
+      success &= findNextBrokerForOosReplica(
+          brokerQueue,
+          unusableBrokers,
+          oosReplica.replicaBrokers,
+          result,
+          oosBrokerId,
+          topicPartition,
+          inBoundReq,
+          outBoundReq,
+          preferredBroker
+      );
+    }
+    // push the brokers back to brokerQueue to keep invariant true
+    brokerQueue.addAll(unusableBrokers);
+    return success ? result : null;
+  }
+
+  /**
+   * Similar to getAlternativeBrokers, but locality aware
+   * @param brokerQueueByLocality a map keeping a priority queue of brokers for each locality
+   * @param oosReplica out of sync replicas
+   * @return a BrokerId to KafkaBroker mapping
+   */
+  public Map<Integer, KafkaBroker> getAlternativeBrokersByLocality(
+      Map<String, PriorityQueue<KafkaBroker>> brokerQueueByLocality,
+      OutOfSyncReplica oosReplica,
+      Map<OutOfSyncReplica, Map<Integer, String>> localityAwareReassignmentFailures
+      ) {
+
+    Map<String, List<Integer>> oosBrokerIdsByLocality = new HashMap<>();
+    for ( int oosBrokerId : oosReplica.outOfSyncBrokers) {
+      String brokerLocality = brokers.get(oosBrokerId).getRackId();
+      oosBrokerIdsByLocality
+          .computeIfAbsent(brokerLocality, l -> new ArrayList<>())
+          .add(oosBrokerId);
+    }
+
+    TopicPartition topicPartition = oosReplica.topicPartition;
```

Comment:
line 322 - 325 is duplicated with line 275-277.  Could we move these few lines to `generateReassignmentPlanForDeadBrokers` and pass the values as parameters to  `kafkaCluster.getAlternativeBrokersByLocality` and `kafkaCluster.getAlternativeBrokers`?

      TopicPartition topicPartition = oosReplica.topicPartition;
      double inBoundReq = ReplicaStatsManager.getMaxBytesIn(zkUrl, topicPartition);
      double outBoundReq = ReplicaStatsManager.getMaxBytesOut(zkUrl, topicPartition);
      int preferredBroker = oosReplica.replicaBrokers.get(0);

      replacedNodes = isLocalityAware
                      ? kafkaCluster.getAlternativeBrokersByLocality(
                          brokerQueueByLocality,
                          oosReplica,
                          reassignmentToLocalityFailures,
                          inBoundReq, outBoundReq, preferredBroker
                        )
                      : kafkaCluster.getAlternativeBrokers(brokerQueue, oosReplica, inBoundReq, outBoundReq, preferredBroker);